src/cc/usdt.cc

```diff
@@ -137,32 +146,79 @@ bool Probe::usdt_cases(std::ostream &stream, const optional<int> &pid) {
   const size_t arg_count = locations_[0].arguments_.size();
 
   for (size_t arg_n = 0; arg_n < arg_count; ++arg_n) {
-    Argument *largest = nullptr;
-    for (Location &location : locations_) {
-      Argument *candidate = location.arguments_[arg_n];
-      if (!largest ||
-          std::abs(candidate->arg_size()) > std::abs(largest->arg_size()))
-        largest = candidate;
-    }
-
-    tfm::format(stream, "%s arg%d = 0;\n", largest->ctype(), arg_n + 1);
+    tfm::format(stream, "%s arg%d = 0;\n", largest_arg_type(arg_n), arg_n + 1);
   }
 
   for (size_t loc_n = 0; loc_n < locations_.size(); ++loc_n) {
     Location &location = locations_[loc_n];
     tfm::format(stream, "if (__loc_id == %d) {\n", loc_n);
 
     for (size_t arg_n = 0; arg_n < location.arguments_.size(); ++arg_n) {
-      Argument *arg = location.arguments_[arg_n];
-      if (!arg->assign_to_local(stream, tfm::format("arg%d", arg_n + 1),
-                                bin_path_, pid))
+      Argument &arg = location.arguments_[arg_n];
+      if (!arg.assign_to_local(stream, tfm::format("arg%d", arg_n + 1),
+                               bin_path_, pid))
         return false;
     }
     stream << "}\n";
   }
   return true;
 }
 
+std::string Probe::largest_arg_type(size_t arg_n) {
+  Argument *largest = nullptr;
+  for (Location &location : locations_) {
+    Argument *candidate = &location.arguments_[arg_n];
+    if (!largest ||
+        std::abs(candidate->arg_size()) > std::abs(largest->arg_size()))
+      largest = candidate;
+  }
+
+  assert(largest);
+  return largest->ctype();
+}
+
+bool Probe::usdt_getarg(std::ostream &stream, const optional<int> &pid) {
+  const size_t arg_count = locations_[0].arguments_.size();
+
+  if (arg_count == 0)
+    return true;
+
+  for (size_t arg_n = 0; arg_n < arg_count; ++arg_n) {
+    std::string ctype = largest_arg_type(arg_n);
+    tfm::format(stream,
+                "static inline %s _bpf_readarg_%s_%d(struct pt_regs *ctx) {\n"
+                "  %s result = 0x0;\n",
+                ctype, name_, arg_n + 1, ctype);
+
+    if (locations_.size() == 1) {
+      Location &location = locations_.front();
+      stream << "  ";
+      if (!location.arguments_[arg_n].assign_to_local(stream, "result",
+                                                      bin_path_, pid))
+        return false;
+      stream << "\n";
+    } else {
+      stream << "  switch(ctx->ip) {\n";
```

Comment:
looks like the loop below will handle different registers at different locations properly. neat approach. I guess the switch(ip) has to stay dynamic unless we can duplicate the program for different ips. The cost of switch is small comparing to the rest. How common is 'locations==1' case?